packages/block-editor/src/components/link-control/index.js

```diff
@@ -173,7 +232,7 @@ function LinkControl( {
 
 		return Promise.resolve( [
 			{
-				id: '-1',
+				id: cleanForSlug( val ),
```

Comment:
Why `cleanForSlug`? I thought it was decided to use the URL (https://github.com/WordPress/gutenberg/pull/19775#discussion_r375334003)? There's been a lot of churn here. I have worry we don't have a strong sense of what we want `id` to represent.

Reply:
I agree. I think this probably changed during my out of hours. I'll need to catch up.

Reply:
That was my doing. I'm using the URL and using the `cleanForSlug` to strip out any potential issues in the user-input URL string that might conflict with the ID being used a react key.

Reply:
For more clarity here on my reasoning for using `cleanForSlug`here.  In this instance, we don’t really care what the ID is, just:
- that there is an ID set
- that it’s a valid react key
- that it’s consistent (testable)

Using whatever the user input as a direct id that would get used to generate a react key (it gets used for this later) felt risky in terms of ensuring it remains a valid react key, so using the existing `cleanForSlug` seemed like a good way to go.

Reply:
We should care what it is, if it's part of the public interface of the component. We're passing these objects through to `onChange`, so we should be able to explain what an `id` represents. Its usage as React key is an implementation detail the consumer shouldn't need to know or care about. The same way good React keys are associated to some known, deterministic, unique property of a data model, we should want be able to point to _what that property is_, conceptually, and for all variations of the suggestion object. I think it's fine if we say that this can be optionally determined based by the type (e.g. a post's ID), with absence of a type deferring to URL as a sensible default (or, again, omitting altogether).

Even as implemented, I'd still have specific technical concerns:

- "felt risky in terms of ensuring it remains a valid react key" - Can we investigate this further, even just to have a good understanding of what makes a valid React key? I wouldn't personally think it should matter what the value is, as long as it's a string.
- Does `cleanForSlug` risk "collisions" for otherwise distinct objects?

Example:

```js
const url1 = 'https://example.com/some.file.or.path';
const url2 = 'https://example.com/some_file_or_path';
console.log( url1 === url2 );
// false
console.log( cleanForSlug( url1 ) === cleanForSlug( url2 ) );
// true
```

Reply:
*tl;dr*: It doesn't appear that there would be an issue with using whatever the user inputs in the field, but is there any concern over having a key constantly updating when we can use a descriptive static key instead?

I'm still new to React, so following all of this is a little tricky, so I appreciate your explanation and patience :)

I took a look again at what these keys are being used for.

<img width="368" alt="Link Control component with manual URL and new page option" src="https://user-images.githubusercontent.com/967608/74274048-0e4f9e80-4cd7-11ea-82eb-d3dae3079c12.png">

1. *The Add URL button*: It will be updated to match whatever you're typing. There's only ever one at a time, and react keys only need to be unique amongst siblings. Rather than updating the `key` on each search, why not just label it `add-url` so it's descriptive and doesn't need to update its key? It's resulting suggestion key would become `add-url-URL`
2. *The Create New Page Button*: It's also updated to match what you're typing, and also only ever has one. We could label this one `create-page`, which ends up becoming `create-page-__CREATE__`

I honestly am not sure about all of this. I did spend a good chunk of time grokking where all this would impact but I'm not sure I found everything. I'm more than happy to go with whatever @aduth and @getdave think is best. This is entirely a learning process for me.

> Can we investigate this further, even just to have a good understanding of what makes a valid React key? I wouldn't personally think it should matter what the value is, as long as it's a string.
It seems like it just needs to be a string, so I guess anything is fine to be used. I tried entering a bunch of different characters as a quick test and did not encounter any errors.

> Does cleanForSlug risk "collisions" for otherwise distinct objects?
In general yes, but in this instance, I don't think so. I'm trying to come up with a scenario where it would conflict, but since it only needs to be unique amongst siblings, I'm not sure it's possible.

Reply:
In terms of `keys` we have the following:

* [Search suggestions coming from an API endpoint use the entity's `id` prop as the key](https://github.com/WordPress/gutenberg/blob/bcd61379396f0e624254dfd200f0d4c8067b5085/packages/block-editor/src/components/link-control/index.js#L465).
* ["Create" suggestion use the "cleaned" version of whatever is typed into the `<input />` plus the string `-__CREATE__`](https://github.com/WordPress/gutenberg/blob/bcd61379396f0e624254dfd200f0d4c8067b5085/packages/block-editor/src/components/link-control/index.js#L447) - it will always be unique amongst siblings because there is only a single "Create" suggestion with that string appended.
* [The "Direct Entry" (URL) suggestion use the "cleaned" version of whatever is typed into the `<input />`plus the string `-URL`](https://github.com/WordPress/gutenberg/blob/bcd61379396f0e624254dfd200f0d4c8067b5085/packages/block-editor/src/components/link-control/index.js#L465) - it will always be unique amongst siblings because there is only a single "Direct Entry" suggestion with that string appended.

Screenshot:
<img width="349" alt="Screenshot 2020-02-13 at 09 32 54" src="https://user-images.githubusercontent.com/444434/74421497-0ada2580-4e45-11ea-9026-21ce2e0fca85.png">


In terms of `id` prop for the suggestion, it's the exact same as the above but with the hardcoded strings removed. This could be an issue with `id` collisions between CREATE and URL types except for the fact that the CREATE option doesn't actually have a suggestion object associated with it:

Screenshot:
<img width="932" alt="Screenshot 2020-02-13 at 09 41 24" src="https://user-images.githubusercontent.com/444434/74421484-06157180-4e45-11ea-8d40-0bd3aa2197fa.png">


Therefore, this seems to satisfy the conditions that:

* `id` should be deterministic
* `id` should be decoupled from the concept of the React key (implementation detail).
* React keys should be unique (ie: avoid "collisions" with otherwise distinct objects)

 The only change we might agree on is what @jeryj suggested: that we choose _not_ to provide an `id` prop to either of the hardcoded suggestion types (CREATE and URL) and instead simply use hardcoded `key`s to ensure they are unique within a collection.

Reply:
So I think there are one of two actions we want to take here?

- It may be as simple as removing `cleanForSlug` here. If the reasons for it to have existed in the first place was to make it into a format that React allows, but we later determined that React will allow anything, then to me it is not needed and can be removed. For me, this can make sense if the `id` is just the URL (or whatever is input). It's not clear to me, but as you're suggesting about the concatenation, I might not like it as much if we're concatenating some magic strings in some cases, or at least wonder if it's necessary. But mostly it would be a worry if it's something that would end up being exposed to the consumer. If it were purely internal, I'd not have as much issue with it.
- Or: "choose not to provide an id prop to either of the hardcoded suggestion types (CREATE and URL) and instead simply use hardcoded keys to ensure they are unique within a collection." I think this is most appropriate if we must expose this to the consumer, and we must have distinction between the "create"-suffixed ID, and we're otherwise unable to come to some consensus on what `id` should "mean". My only worry here is how well we're accounting for suggestions which include only some potential properties of a value (#20020).

Especially in light of the previous discussion of how we're the documentation of "suggestion can be used for the next value", I'd want to be sure that we don't have situations where a consuming component (or worse, persisted block attributes) is storing a value like `{ id: 'http-example-__CREATE__' }`.

Reply:
> It may be as simple as removing cleanForSlug here. If the reasons for it to have existed in the first place was to make it into a format that React allows, but we later determined that React will allow anything

I don't think I've seen any docs to 100% prove this but 🤷‍♂ 

[The docs](https://reactjs.org/docs/reconciliation.html#keys), do say a `key` is used to reconcile differences between two "trees". Therefore "[The key only has to be unique among its siblings, not globally unique](https://reactjs.org/docs/lists-and-keys.html#keys)".

> Or: "choose not to provide an id prop to either of the hardcoded suggestion types (CREATE and URL) and instead simply use hardcoded keys to ensure they are unique within a collection." I think this is most appropriate if we must expose this to the consumer, and we must have distinction between the "create"-suffixed ID, and we're otherwise unable to come to some consensus on what id should "mean". My only worry here is how well we're accounting for suggestions which include only some potential properties of a value (#20020).

I'm not sure we can guarantee this will be ok. The "CREATE" option does not pass the `id` prop to its consumer. In fact the only prop it exposes is `suggestion.title` as an arg when calling the `createSuggestion` prop. Therefore we don't _need_ the `id` prop and we can hardcode the React key.

However, the "Direct Entry" (aka "URL") option exposes the entire `suggestion` object because it is passed to the `onChange` handler prop. Therefore it needs to be something that is unique and valid.

I'd always seen the `id` as completely unique value that identifies a given suggestion. For `Post` objects that's easy - the Post Id. But for the "Direct Entry" suggestion we only have what the user types into the input field. Therefore what uniquely identifies the resulting suggestion is in fact what the user typed into the input field. In this case because there can only ever be a single "Direct Entry" result, we can be sure that the URL will uniquely identify the record. In terms of React `key` we can simply hardcode this as again there is only ever 1 "Direct Entry" suggesiton.

One final thought might be using something like `btoa` to produce a "hash" like representation of the inputted `url` which might like more like a traditional `id`, but honestly I'm struggling to see what value that would bring.

In summary, I'm leaning towards:

1. Removing `cleanForSlug`.
2. Using inputted URL value as the `id` prop for "Direct Entry" types.
3. Removing the `id` from "Create" types entirely.
4. Using a hardcoded React `key` value for CREATE.
5. Using the URL string plus the hardcoded string `URL` as the React `key` for "Direct Entry" types.

Note that for 5 above we need to continue to use concatenation for the React `key` (only - not for the `id`) because for normal API search results, we can't rely on the `id` always being unique amongst siblings because we might mix different types of Entity into the search results. Say for example we started returning Categories as well as Pages in the results. We'd have clashing `id`s because Post and Categories are from different tables and may have the same IDs. Therefore appending the `type` to the end of the React key avoids this.

Reply:
> I'd always seen the `id` as completely unique value that identifies a given suggestion. For `Post` objects that's easy - the Post Id. But for the "Direct Entry" suggestion we only have what the user types into the input field. Therefore what uniquely identifies the resulting suggestion is in fact what the user typed into the input field. In this case because there can only ever be a single "Direct Entry" result, we can be sure that the URL will uniquely identify the record. In terms of React `key` we can simply hardcode this as again there is only ever 1 "Direct Entry" suggesiton.

This explanation makes sense to me. I can grant it might be odd that the `id` of a suggestion would be some partially-crafted user input, but like you note, this is the most accurate way to consider its uniqueness. I don't think any further processing (`cleanForSlug`, `btoa`) really does much for us rather than perhaps conform better to what we imagine an `id` to look like. But unless we can be very clear about what that is, I don't think we ought to bother.

>In summary, I'm leaning towards:

This all sounds fine to me.

>Note that for 5 above we need to continue to use concatenation for the React `key`

It's a very small overhead in the render, but it does not trouble me much otherwise, as long as it allows us to retain what we want the suggestion (and its properties) to represent, while still allowing us to satisfy the implementation requirements imposed on us by React.

Reply:
@aduth As agreed I'm working through these:

- [ ] Removing cleanForSlug.
- [ ] Using inputted URL value as the id prop for "Direct Entry" types.
- [ ] Removing the id from "Create" types entirely.
- [ ] Using a hardcoded React key value for CREATE.
- [ ] Using the URL string plus the hardcoded string URL as the React key for "Direct Entry" types.

Reply:
Here are the updated results showing `key` and `id` (where present):

------

### Direct Entry
<img width="1136" alt="Screenshot 2020-02-14 at 15 44 43" src="https://user-images.githubusercontent.com/444434/74545640-1f9be380-4f41-11ea-8b21-66a765176bbf.png">

* `key`: `https://directentryurl.com`
* `id`: `https://directentryurl.com`
* Unique? Yes as only ever 1 Direct entry result.

------

### "Create"

<img width="1126" alt="Screenshot 2020-02-14 at 15 44 15" src="https://user-images.githubusercontent.com/444434/74545671-2fb3c300-4f41-11ea-892a-fb4876a831f0.png">

* `key`: `__CREATE__`
* `id`: n/a (not present)
* Unique? Yes as only ever 1 Create result.

------
### Standard Search Result

<img width="1014" alt="Screenshot 2020-02-14 at 15 45 23" src="https://user-images.githubusercontent.com/444434/74545689-3b06ee80-4f41-11ea-8303-562401f71c9c.png">

* `key`: `15-page` (Post ID + Post Type)
* `id`: `15` (just Post ID)
* Unique? Highly likely as even with search records coming from separate DB tables using both `id` and `type` helps to avoid clashes which could happen if you relied on ID only.